### ISSUE TITLE
Need clarification regarding the relationship between Activation Functions and the Initializers

### ISSUE THREAD

**Issue by worldpeaceaspirer** [ageron/handson-ml2#437]
In the **`Table 11-1`**, shown below, the formula for **`Glorot`** and **`He Initializers`** is dependent on the **`Activation Functions`**.

But in the Tensorflow Source Code of neither the [He Initializer](https://github.com/tensorflow/tensorflow/blob/a4dfb8d1a71385bd6d122e4f27f86dcebb96712d/tensorflow/python/keras/initializers/initializers_v2.py#L849-L883), nor the [Glorot Initializer](https://github.com/tensorflow/tensorflow/blob/a4dfb8d1a71385bd6d122e4f27f86dcebb96712d/tensorflow/python/keras/initializers/initializers_v2.py#L711-L752), along with the Code of the Parent Class, [VarianceScaling Initializer](https://github.com/tensorflow/tensorflow/blob/v2.5.0/tensorflow/python/keras/initializers/initializers_v2.py#L430-L531), there is no mention of **`Activation Function's`** dependency on the **`Initializers`**.

This should be either a mistake in the book or should be a bug in the Tensorflow Code. Can you please clarify?

![image](https://user-images.githubusercontent.com/48206667/120094668-66661a80-c13f-11eb-8398-8c381f56e182.png)

**Image from the first version of "Hands on Machine Learning with Scikit-Learn and Tensorflow":**

![image](https://user-images.githubusercontent.com/48206667/120095045-4a637880-c141-11eb-9550-c4524bb728d5.png)

**Comment by ageron** [comment]
Hi @worldpeaceaspirer ,

That's a great question! Here are some reasons that might explain why the main deep learning libraries do not try to automatically match the initializer with the activation function:

* firstly, whether you choose to use Glorot initialization or He initialization makes little difference in many neural nets (especially shallow ones). That said, it does make a difference for deeper nets (or else I wouldn't have mentioned it in the book), so this argument is insufficient.
* secondly, it's often best for an API to be as simple and consistent as possible, even if it means not always being optimal. Adding too much magic might do more harm than good. Having one default initializer is simple and easy to understand, even if it is not always optimal. If you're training a deep net, there are things you will need to tweak to get the best results, including the initializers.
* thirdly, and perhaps most importantly, it might be quite hard to implement. Indeed, what about neural nets where the layer containing the weights is not directly attached to the activation function (for example, there may be a BatchNormalization layer in between a Convolutional layer and the activation function)? In PyTorch, the layers with weights and the activation function are treated separately, so it would be very tricky to tie them automatically. And what about user-provided activation functions? Should the user be forced to provide both the activation function and the corresponding initializer? If not, then what should the default initializer be?

Note that Keras uses the Glorot uniform initialization by default, while PyTorch uses He (Kaiming) uniform initialization by default. Keras lets you easily choose another initializer when creating a layer, by setting `kernel_initializer`. In PyTorch, you have to set the weights yourself after creating the layer, typically using one of the provided initialization functions:

```python
dense = torch.nn.Linear(10, 3) # 10 inputs, 3 outputs
torch.nn.init.xavier_uniform_(dense.weight)
```

Interestingly, PyTorch provides a `calculate_gain()` function that returns the "gain" as a function of the initialization function. The gain is the number by which the Glorot initialization must be multiplied. For example, it returns sqrt(2) for ReLU. It also supports some more complex activation functions, such as Leaky ReLU, for example:

```python
gain = torch.nn.init.calculate_gain("leaky_relu", 0.2)  # leaky_relu with negative_slope=0.2
torch.nn.init.xavier_uniform_(dense.weight, gain=gain)
```

I hope this helps!

**Comment by ageron** [comment]
Note that I recommended a gain of 2 for the hyperbolic activation function (𝑡𝑎𝑛ℎ) in the first edition of my book (I don't remember what source I got that from, but it seemed reasonable given that 𝑡𝑎𝑛ℎ(𝑥)=2⋅𝜎(2𝑥)−1). However, the paper [On weight initialization in deep neural networks](https://arxiv.org/abs/1704.08863) by Siddharth Krishna Kumar recommended using a gain of 1, which explains the change in the second edition.

However, looking at the `calculate_gain()` function a few minutes ago, I was surprised to see that it returns 5/3 for 𝑡𝑎𝑛ℎ! I could not find a paper to support this number, but Thomas Viehmann gave an interesting justification in [this discussion](https://discuss.pytorch.org/t/calculate-gain-tanh/20854).

Similarly, the authors of the SELU paper recommend using a gain of 1, but the `calculate_gain()` function returns 3/4 instead. The discussion above also explains why they use this value.

**Comment by worldpeaceaspirer** [comment]
@ageron,
Sorry for the delayed response. It's **my pleasure discussing with you because I have learnt many concepts of Machine Learning and Deep Learning from your books**. Thank you for the detailed explanation. Closing the issue.
